### DISCUSSION TITLE
How can I add operators to `formatQuery(query, 'jsonlogic')`

### DISCUSSION THREAD

**Discussion by tomercohen1210** [react-querybuilder/react-querybuilder#376] [Q&A] [answered]
Hey, I am trying to add new operators to `formatQuery(query, 'jsonlogic')`, it keeps returning `false` if I use a new operator.
Can you help me with how?
Thanks!

**Comment by jakeboone02** [comment]
Hi @tomercohen1210. What would you like to add? A code sample would be really helpful so I can see exactly what you're trying to accomplish.

**Comment by tomercohen1210** [comment]
Hi @jakeboone02 , I am trying to parse the query with my operator but keep getting `false` from `formatQuery(query, 'jsonlogic')`
for example :
`{
    "combinator": "and",
    "rules": [
        {
            "field": "coverage_cycle",
            "operator": "have",
            "value": "P1"
        }
    ]
}`
how can I add to the `formatQuery` new operator so it won't return `false`?
thanks!

**Comment by jakeboone02** [comment]
`formatQuery` only handles the operators in the type `DefaultOperatorName`, implemented in the export `defaultOperators`, and "have" isn't in that list.

My advice would be to map any non-default operators to something in the default list and run your query through [`transformQuery`](https://react-querybuilder.js.org/docs/api/misc#transformquery) prior to `formatQuery`, like this:

```ts
formatQuery(
  transformQuery(query, { operatorMap: { have: 'in' } }),
  'jsonlogic'
)
```

(That's assuming "have" is equivalent to "in". If it's more like "contains" then use that instead.)

**Reply by tomercohen1210** [reply]
@jakeboone02 is there any way to add to `defaultOperators`?
I did try `contains`, but the `formatQuery` function transfer it to `in`, why Is that?

**Reply by jakeboone02** [reply]
The "in" operation in JsonLogic has two meanings depending on the associated value.

The [array operation](https://jsonlogic.com/operations.html#in) means "this item is in this list of items." It looks like this:

```json
{"in": ["Ringo", ["John", "Paul", "George", "Ringo"] ]}
```

The [string operation](https://jsonlogic.com/operations.html#in-1) means "this string is a subset of this other string." It looks like this:

```json
{"in": ["Spring", "Springfield"]}
```

`formatQuery` converts "in" into the array operation and "contains" into the string operation.

**Comment by tomercohen1210** [comment]
@jakeboone02 is there any why that i can tell the function to no change `contains` to `in`?

**Reply by jakeboone02** [reply]
No, would you want to change it to something else?

**Reply by tomercohen1210** [reply]
To be honest, I will expect the `foramtQuery` function to be the opposite of the `parseJsonLogic`.
`parseJsonLogic` get a jsonLogic and just change its format to be acceptable by the `Query`, so it makes sense that `foramtQuery` will do the opposite and just change the format to jsonLogic without checking the operators or changing them.
Currently:
`parseJsonLogic` with - 
```
{
  "have": [
    {
      "var": "firstName"
    },
    "Tomer"
  ]
}
```
returns
```
{
  "rules": [
    {
      "field": "firstName",
      "value": "Tomer",
      "operator": "have"
    }
  ],
  "combinator": "and",
  "not": false
}
```
But if I put the result in `foramtQuery` I get `false` and this just doesn't make sense to me, I will expect to get the same as at the start.
@jakeboone02 What do you think?

**Reply by jakeboone02** [reply]
Yeah ok, I see what you're saying now. We actually have a couple of additional operators that are supposed to be added with `jsonLogic.add_operation` before applying the result of `formatQuery`, although I'm just now noticing that they're not documented other than a [comment in the code](https://github.com/react-querybuilder/react-querybuilder/blob/0487419f0603b19d32c2eb00d6be25bf77922d86/packages/react-querybuilder/src/utils/formatQuery/utils.ts#L49-L65). I'll fix that soon!

We could probably enable custom operators in `formatQuery`, but I'm just wondering how it would know how to handle the values. For example, the "in" operator becomes `{ "in": [field, [value1, value2, ..., valueN]] }`, but the "contains" operator becomes `{ "in": [field, value] }`. How should a custom operator be defined so that the value gets constructed correctly? Would there be a template of some kind?

**Reply by tomercohen1210** [reply]
About the `jsonLogic.add_operation` I saw the comment, but I didn't understand from where I can Import `jsonLogic` do I would be able to do this action, If you can show me some examples? 

I think the best for a custom can be something like this `{ have: (field, value) => <how the user want to get it> }`
what do you think?

**Reply by jakeboone02** [reply]
For the JavaScript implementation, get `jsonLogic` from the [`json-logic-js`](https://www.npmjs.com/package/json-logic-js) package. I assumed that's what you were already using.

```ts
import jsonLogic from "json-logic-js";
// ...
```

As for the function "template" you've got the right idea. I'm going to add a `ruleProcessor` function to the existing `formatQuery` options. The rule processor would supersede the default rule processor but could fall back to the default if necessary. PR will be linked here soon.

**Reply by jakeboone02** [reply]
See #378.

**Reply by tomercohen1210** [reply]
@jakeboone02 wow man thanks!
Can you show me an example how to use it for `{ have: (field, value) => [field, value] }`?
Is it like this? 
when are you going to release it for 5.alpha?
```
foramtQuery(query, {
fomat: 'jsonlogic',
ruleProcessor: { have: (field, value) => [field, value] }
)
```

**Reply by jakeboone02** [reply, marked as answer]
Sure, see here: https://react-querybuilder.js.org/docs/next/api/export#rule-processor

**Reply by tomercohen1210** [reply]
Is it in `v5.0.0-alpha.6`?

**Reply by jakeboone02** [reply]
No, 5.0.0-alpha.7. Released last night.
